src/devices/machine/amiga_copper.cpp

```diff
@@ -0,0 +1,397 @@
+{
+	int word0, word1;
+
+	/* bail if not enabled */
+	if (!m_dma_master_enable || !m_dma_copen)
+		return 511;
+
+	/* flush any pending writes */
+	if (m_pending_offset)
+	{
+		//LOGCHIPSET("%02X.%02X: Write to %s = %04x\n", ypos, xpos / 2, s_custom_reg_names[m_copper_pending_offset & 0xff], m_copper_pending_data);
+		LOGCHIPSET("%02X.%02X: Write to %s = %04x\n",
+			ypos,
+			xpos / 2,
+			m_pending_offset << 1,
+			m_pending_data
+		);
+		m_host_space->write_word(0xdff000 | (m_pending_offset << 1), m_pending_data);
+		m_pending_offset = 0;
+	}
+
+	/* if we're waiting, check for a breakthrough */
+	if (m_state_waiting)
+	{
+		int curpos = (ypos << 8) | (xpos >> 1);
+
+		/* if we're past the wait time, stop it and hold up 2 cycles */
+		if ((curpos & m_waitmask) >= (m_waitval & m_waitmask) &&
+			(!m_state_waitblit || !(is_blitter_busy)))
+		{
+			m_state_waiting = false;
+//#if GUESS_COPPER_OFFSET
+//			return xpos + COPPER_CYCLES_TO_PIXELS(1 + m_wait_offset);
+//#else
+			return xpos + COPPER_CYCLES_TO_PIXELS(1 + 3);
+//#endif
+		}
+
+		/* otherwise, see if this line is even a possibility; if not, punt */
+		if (((curpos | 0xff) & m_waitmask) < (m_waitval & m_waitmask))
+			return 511;
+
+		/* else just advance another pixel */
+		xpos += COPPER_CYCLES_TO_PIXELS(1);
+		return xpos;
+	}
+
+	/* fetch the first data word */
+	// TODO: swap between ir0 and ir1 is controlled thru a selins latch
+	// (which can't be this instant too)
+	word0 = m_chipmem_r(m_pc);
+	m_host_space->write_word(0xdff08c, word0);
+	m_pc += 2;
+	xpos += COPPER_CYCLES_TO_PIXELS(1);
+	
+	/* fetch the second data word */
+	word1 = m_chipmem_r(m_pc);
+	m_host_space->write_word(0xdff08c, word1);
+	m_pc += 2;
+	xpos += COPPER_CYCLES_TO_PIXELS(1);
+
+	LOGINST("%02X.%02X: Copper inst @ %06x = %04x %04x\n",
+		ypos,
+		xpos / 2,
+		m_pc - 4,
+		word0,
+		word1
+	);
+
+	/* handle a move */
+	if ((word0 & 1) == 0)
+	{
+		/* do the write if we're allowed */
+		word0 = (word0 >> 1) & 0xff;
+		if (word0 >= m_cdang_setting)
+		{
+			if (delay[word0] == 0)
+			{
+				//LOGCHIPSET("%02X.%02X: Write to %s = %04x\n", ypos, xpos / 2, s_custom_reg_names[word0 & 0xff], word1);
+				LOGCHIPSET("%02X.%02X: Write to %s = %04x\n",
+					ypos,
+					xpos / 2,
+					word0 << 1,
+					word1
+				);
+				m_host_space->write_word(0xdff000 | (word0 << 1), word1);
+			}
+			else    // additional 2 cycles needed for non-Agnus registers
+			{
+				m_pending_offset = word0;
+				m_pending_data = word1;
+			}
+		}
+
+		/* illegal writes suspend until next frame */
+		else
+		{
+			LOGWARN("%02X.%02X: Aborting copper on illegal write\n", ypos, xpos / 2);
+
+			m_waitval = 0xffff;
+			m_waitmask = 0xffff;
+			m_state_waitblit = false;
+			m_state_waiting = true;
+
+			return 511;
+		}
+	}
+	else
+	{
+		/* extract common wait/skip values */
+		m_waitval = word0 & 0xfffe;
+
+		// ???
+		#ifdef UNUSED_FUNCTION
+		if (m_waitval != 0xfffe)
+			m_waitval = (word0 & 0x00fe) | ((((word0 >> 8) & 0xff) + 1) << 8);
```

Comment:
This was originally a `#if 0`, gotta decide if worth keeping (looks just a debug left-over not worth having it)